core/src/main/scala/chisel3/properties/Property.scala

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3.properties
+
+import chisel3.{Bits, Element, PString, Printable, UInt}
+import chisel3.experimental.SourceInfo
+import chisel3.internal.throwException
+import chisel3.internal.firrtl.Width
+import scala.reflect.runtime.universe.{typeOf, TypeTag}
+import scala.annotation.implicitNotFound
+
+/** PropertyType defines a typeclass for valid Property types.
+  *
+  * Typeclass instances will be defined for Scala types that can be used as
+  * properties. This includes builtin Scala types as well as types defined in
+  * Chisel.
+  */
+@implicitNotFound("unsupported Property type ${T}")
+private[chisel3] class PropertyType[T]
+
+/** Companion object for PropertyType.
+  *
+  * Typeclass instances for valid Property types are defined here, so they will
+  * be in the implicit scope and available for users.
+  */
+private[chisel3] object PropertyType {}
+
+/** Property is the base type for all properties.
+  *
+  * Properties are like normal Elements in that they can be used in ports,
+  * connected to other properties, etc. However, they are used to describe a set
+  * of non-hardware types, so they have no width, cannot be used in hardware
+  * aggregates, and cannot be connected to hardware types.
+  */
+class Property[T: PropertyType: TypeTag] extends Element {
```

Comment:
I don't think that `Property` should extend `Element` because other than the use in ports, they are very much not sharing much in common with the other `Elements`, (as evidenced by all of the overridden methods that just throw exceptions). I get that you need to figure out how to make them be allowed in Ports, but I think maybe we need a split hierarchy here, I don't think Properties are `Data`, they are a new thing.

Reply:
Thanks, this was a question I had, and your reasoning makes sense to me. I think cordoning these off will be great. The only real functionality I do want from `Data` was `bind` and the nice `:=` syntax, which I guess we will just implement this separately for properties when we get to being able to assign them to outputs, etc. Looking a bit ahead, there will be some other laundry list items to make work with properties, like BoringUtils. But I don't think making them their own hierarchy makes that any harder.

Reply:
I am taking a look at what this might look like. It seems like we could almost update `Port` so it's `id` field is `HasId` instead of Data. There are still some things to figure out how to update, but then we could make `Property` extend `HasId`. Or do you think there is a better way to attack this?

Reply:
As I pull on this thread, I guess what we want is roughly a trait that extends `HasId`, has a specified direction, and has some of the clonetype machinery. Maybe some other stuff I haven't uncovered. And both Data and Property would extend this trait. Does that seem about right?

Reply:
since you mentioned it, do you expect Connectables to work on Property aggregates? Like propertyIn :<>= propertyOut? It seems like we'd need to figure out if that is really the same thing at all or just something that happens to look the same but is defined on your property type.

By the way, i know you don't want properties inside aggregates in firrtl, but we might consider whether it makes sense to have them inside chisel3.Bundle, and thus doing :<>= on a bundle that contains properties would also assign the properties...

Reply:
> just something that happens to look the same but is defined on your property type.

I haven't gotten to aggregates in my prototyping, but i guess in this world that is what it would look like: similar surface syntax to Chisel users, but implemented separately for Properties.

In my original prototype that this PR is split from, when I went with the notion that a Property is Data, it was fairly straightforward to add support for Property elements in MonoConnect, and I guess it would be similar for BulkConnect.

But I think the suggestion that Property is not Data makes sense, and we can set up a convenient user experience that feels like you're connecting Properties like you'd connect Data.

Reply:
I think it's easy enough to just implement the methods we want rather than trying to reuse the machinery of Data, largely because the machinery of Data is not implemented very well and has a lot of issues. We have an opportunity here to create similar feeling APIs that work and compose better and then perhaps apply some of those learnings back to Data.

Reply:
I *think* `cloneIORecord` is only used for `CloneModuleAsRecord` which I think we just forgot to deprecate as it was sort of a proto-D/I API that we really don't need to support. For purposes of Properties, just ignore it and separately I think we should deprecate it.

Reply:
Ok, the refactoring to add a new BaseType has merged, so we can resolve this discussion.